File: C#/src/Splitit.SDK.Client/Client/FlexFields.cs
using Splitit.SDK.Client.Model;
using Splitit.SDK.Client.Api;
using System.Collections.Generic;
using System;

namespace Splitit.SDK.Client.Client
{
    public class FlexFields
    {
        private InstallmentPlanApi _api;
        private Configuration _env;

        private InitiateInstallmentPlanRequest _request = new InitiateInstallmentPlanRequest(){
            PlanData = new PlanData()
        };


        public static FlexFields Authenticate(Configuration env, string username, string password)
        {
            var ff = new FlexFields();
            ff._env = env;
            ff._api = new InstallmentPlanApi(env, sessionId: ff.GetSessionId(username, password));
            return ff;
        }
        public static FlexFields Authenticate(string apiKey, string username, string password, bool sandbox = false)
        {
            var env = sandbox ? Configuration.Sandbox : Configuration.Default;
            env.AddApiKey(apiKey);

            var ff = new FlexFields();
            ff._env = env;
            ff._api = new InstallmentPlanApi(env, sessionId: ff.GetSessionId(username, password));
            return ff;
        }

        public FlexFields AddCulture(string culture)
        {
            this._api.Culture = culture;
            return this;
        }

        public FlexFields AddInstallments(List<int> installmentOptions, int? defaultNumInstallments = null)
        {
            this._request.PlanData.NumberOfInstallments = defaultNumInstallments;

            if (this._request.PaymentWizardData == null)
            {
                this._request.PaymentWizardData = new PaymentWizardData();
            }

            this._request.PaymentWizardData.RequestedNumberOfInstallments = string.Join(",", installmentOptions);
            return this;
        }

        public FlexFields AddBillingInformation(AddressData addressData = null, ConsumerData consumerData = null)
        {
            this._request.BillingAddress = addressData;
            this._request.ConsumerData = consumerData;
            return this;
        }

        public FlexFields Add3DSecure(RedirectUrls redirectUrls)
        {
            this._request.PlanData.Attempt3DSecure = true;
            this._request.RedirectUrls = redirectUrls;
            return this;
        }

        public FlexFields AddCaptureSettings(
            bool? autoCapture = null, 
            decimal? firstInstallmentAmount = null,
            string currencyCode = null,
            DateTime? firstChargeDate = null)
        {
            this._request.PlanData.AutoCapture = autoCapture;
            
            if (firstInstallmentAmount != null){
                this._request.PlanData.FirstInstallmentAmount = new MoneyWithCurrencyCode(firstInstallmentAmount.Value, currencyCode);
            }
            
            this._request.PlanData.FirstChargeDate = firstChargeDate;
            return this;
        }

        public string GetPublicToken(decimal amount, string currencyCode)
        {
            this._request.PlanData.Amount = new MoneyWithCurrencyCode(amount, currencyCode);

            var initResponse = this._api.InstallmentPlanInitiate(this._request);
            return initResponse.PublicToken;
        }

        public bool VerifyPayment(string planNumber, decimal orderAmount)
        {
            var paymentVerificationResponse = this._api.InstallmentPlanVerifyPayment(new VerifyPaymentRequest(planNumber));
            if (paymentVerificationResponse.IsPaid == true && paymentVerificationResponse.OriginalAmountPaid == orderAmount)
            {
                return true;
            }

            return false;
        }

        private string GetSessionId(string username, string pass)
        {
            var loginApi = new LoginApi(this._env);
            var request = new LoginRequest(userName: username, password: pass);

            var loginResult = loginApi.LoginPost(request);
            return loginResult.SessionId;
            
        }
    }
}

File: python/splitit/flex_fields.py
import splitit


class FlexFields(object):
    def __init__(self, api_client):
        self.api_client = api_client
        self._active_init_req = splitit.InitiateInstallmentPlanRequest(
            plan_data=splitit.PlanData()
        )

    def add_installments(self, installment_options, default_num_installments):
        self._active_init_req.plan_data.number_of_installments = default_num_installments
        if self._active_init_req.payment_wizard_data is None:
            self._active_init_req.payment_wizard_data = splitit.PaymentWizardData()

        self._active_init_req.payment_wizard_data.requested_number_of_installments = installment_options

    def add_billing_information(self, billing_address, consumer_data):
        self._active_init_req.billing_address = billing_address
        self._active_init_req.consumer_data = consumer_data

    def get_public_token(self, amount, currency_code):
        self._active_init_req.plan_data.amount = splitit.MoneyWithCurrencyCode(
            amount, currency_code
        )

        resp = self.api_client.InstallmentPlanApi.installment_plan_initiate(
            self._active_init_req
        )

        return resp.public_token

    def verify_payment(self, plan_number, order_amount):
        resp = self.api_client.InstallmentPlanApi.installment_plan_verify_payment(
            splitit.VerifyPaymentRequest(
                plan_number
            )
        )
        return resp.is_paid and resp.original_amount_paid == order_amount

    def add_3D_secure(self, redirect_urls):
        self._active_init_req.plan_data.attempt3_d_secure = True
        self._active_init_req.redirect_urls = redirect_urls

    def add_capture_settings(self, auto_capture=False, first_installment_amount=0, first_charge_date=None):
        self._active_init_req.plan_data.auto_capture = auto_capture
        if first_installment_amount > 0:
            self._active_init_req.plan_data.first_installment_amount = first_installment_amount
        if first_charge_date:
            self._active_init_req.plan_data.first_charge_date = first_charge_date
